Allow named parameters in mapping types

zemse opened the pull request:
Closes #11407

- Allows providing an optional name/identifier just after the key type and the value type.
- Identifier defaults to empty string "".
- The optional identifier if provided is set to the "name" field in JSON ABI.
- Test cases are added for parser and ABI generator

Please let me know of any changes that are needed.

zemse committed:
feat: allow named parameters in mapping types

Co-authored-by: Hari <webmail.hari@gmail.com>

test: add parser and abi test cases

docs: add example on using named parameters for mappings

- Add changelog

feat: update antlr grammar to allow named parameters in mappings

fix: prevent conflicting mapping parameter names

ref: change order of mapping initializers

test: update expectations and fix build

test: add more tests

fix: use common error & code for conflicting params

fix: issue with accessing nested mapping

test: add conflicting params tests for more nested levels

Update libsolidity/analysis/DeclarationTypeChecker.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

fix: error reported with the same code twice

test: add more tests for 3 level nested mapping

Address review comments

zemse commented on test/libsolidity/errorRecoveryTests/do_not_delete_at_error.sol:
```diff
@@ -10,4 +10,6 @@ contract Error2 {
 }
 // ----
-// ParserError 6635: (417-425): Expected ')' but got identifier
+// ParserError 6635: (425-426): Expected ')' but got ';'
```
Some existing error recovery tests change their behavior because a syntax that was illegal before (i.e. ability to place an identifier next to type in mapping) is legal now, hence the parser passes further and generates a different error.

hrkrshnn commented on libsolidity/ast/AST.h:
```diff
@@ -1428,18 +1428,24 @@ class Mapping: public TypeName
+		ASTPointer<ASTString> _valueName
 	):
-		TypeName(_id, _location), m_keyType(std::move(_keyType)), m_valueType(std::move(_valueType)) {}
+		TypeName(_id, _location), m_keyType(std::move(_keyType)), m_valueType(std::move(_valueType)), m_keyName(std::move(_keyName)), m_valueName(std::move(_valueName)) {}
```
Nit: can you make a new line?

hrkrshnn commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -1183,11 +1183,21 @@ ASTPointer<Mapping> Parser::parseMapping()
 	else
 		fatalParserError(1005_error, "Expected elementary type name or identifier for mapping key type");
+	ASTPointer<ASTString> keyName;
+	if(m_scanner->currentToken() == Token::Identifier)
```
```suggestion
	if (m_scanner->currentToken() == Token::Identifier)
```

hrkrshnn commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -1183,11 +1183,21 @@ ASTPointer<Mapping> Parser::parseMapping()
 	expectToken(Token::DoubleArrow);
 	ASTPointer<TypeName> valueType = parseTypeName();
+	ASTPointer<ASTString> valueName;
+	if(m_scanner->currentToken() == Token::Identifier)
```
```suggestion
	if (m_scanner->currentToken() == Token::Identifier)
```

hrkrshnn commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -1183,11 +1183,21 @@ ASTPointer<Mapping> Parser::parseMapping()
+	ASTPointer<ASTString> keyName;
+	if(m_scanner->currentToken() == Token::Identifier)
+		keyName = getLiteralAndAdvance();
+	else
+		keyName = make_shared<ASTString>("");
```
Can you add a test for something like `mapping (uint uint => uint)`. Should be a parser error.

hrkrshnn commented on test/libsolidity/ABIJson/mapping.sol:
```diff
@@ -0,0 +1,33 @@
+contract test {
+    mapping(address owner => mapping(address spender => uint value)) public allowance;
```
Can you add one more mapping, but with some types that are not annotated?

hrkrshnn replied:
I'll add this now

hrkrshnn reviewed (commented):
Looks quite good! Some small comments.

We were wondering with the annotations, if it makes sense to have the function call options syntax for mappings. For example, for `mapping (address user => uint balance) balanceOf`, would the syntax `balanceOf[{user: 0xaddress}]` be allowed? It's not needed to implement it in this PR, but worth thinking about.

Also, it's missing a changelog entry.

hrkrshnn commented on test/libsolidity/syntaxTests/parsing/mapping_with_names_5.sol:
```diff
@@ -0,0 +1,4 @@
+contract test {
+    mapping(address owner => mapping(address spender => bytes32[] notes)) names;
```
Can you add a test where these symbols conflict with an existing definition? As an example add a `address owner` as a state variable.

hrkrshnn commented:
Can you add a semantic test as well? https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/getters/mapping.sol here is an example

hrkrshnn commented:
Could you also add an example in the documentation?

zemse commented:
Thanks for the review! I've made the changes.

> function call options syntax for mappings

This might be tricky for nested mappings: `allowances[{owner: 0xAddr}][{spender: 0xAddr}]`. Somehow having it together could be good `allowances[{owner: 0xAddr, spender: 0xAddr}]`, however the names are optional hence this together syntax might not work. But yeah having it makes sense for improving code readability, does not affect any ABI or bytecode. I'd prefer adding it in a separate PR unless it is needed in this PR.

Marenz commented on docs/types/mapping-types.rst:
```diff
@@ -4,12 +4,13 @@
 
-Mapping types use the syntax ``mapping(KeyType => ValueType)`` and variables
-of mapping type are declared using the syntax ``mapping(KeyType => ValueType) VariableName``.
+Mapping types use the syntax ``mapping(KeyType KeyName => ValueType ValueName)`` and variables
```
It might be nice to see both ways written out here, if one is not carefully reading, they might think this is not optional (the optional part is hidden in the full text rather than easily visible at a glance)

zemse replied:
I've added a commit which mentions that `mapping(KeyType => ValueType)` also works.

Marenz commented on libsolidity/parsing/Parser.cpp:
```diff
@@ -1183,11 +1183,21 @@ ASTPointer<Mapping> Parser::parseMapping()
+	if (m_scanner->currentToken() == Token::Identifier)
+		keyName = getLiteralAndAdvance();
+	else
+		keyName = make_shared<ASTString>("");
```
You create an `ASTString` here, but in most other places you use `std::string`. 
This is technically fine as `ASTString` is just an alias currently, but for consistency and to avoid future problems if we decide to change that alias, `ASTString` should be used everywhere :)

zemse replied:
Thanks for pointing this out. I've replaced all `std::string` with `ASTString`.

Marenz reviewed (changes requested):
already looking very good.
Additionally to the comments, I am wondering if it makes sense to make it an `optional<ASTPointer<ASTString>>`. 
Would love to hear @cameel and @christianparpart on that topic

zemse commented:
I have just made the review changes requested and rebased over the latest develop branch.

> Depends if we prefer to treat it as a missing name or an empty name. 

I think if names are made optional, that'd need a small condition while generating ABI because if name is missing, then it has to create an empty string "" anyway for pushing it to param names array. But do lmk if this change is needed to be done in this PR.

zemse commented:
I have rebased over the latest `develop` branch to fix the conflicts.

hrkrshnn commented:
Will rebase now

hrkrshnn commented on Changelog.md:
```diff
@@ -1,6 +1,7 @@
 ### 0.8.18 (unreleased)
 
 Language Features:
+* Allow mapping key and value identifiers
```
```suggestion
* Allow named parameters in mapping types.
```

hrkrshnn commented on docs/types/mapping-types.rst:
```diff
@@ -64,6 +67,25 @@ contract that returns the value at the specified address.
+.. code-block:: solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >=0.8.17;
```
```suggestion
    pragma solidity ^0.8.17;
```
We like to restrict the pragmas in the documentation. Mainly because the documentation will still be available and there is no guarantee that the code with `>= 0.8.17` will compile on `0.9.0` for example.

hrkrshnn commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -269,12 +269,15 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
 	// Convert value type to storage reference.
 	valueType = TypeProvider::withLocationIfReference(DataLocation::Storage, valueType);
-	_mapping.annotation().type = TypeProvider::mapping(keyType, valueType);
+	_mapping.annotation().type = TypeProvider::mapping(keyType, valueType, keyName, valueName);
```
Nit: I would have preferred the order to be `keyType, KeyName, valueType, valueName`. But no strong argument on changing this.

hrkrshnn commented on libsolidity/ast/AST.h:
```diff
@@ -1428,18 +1428,29 @@ class Mapping: public TypeName
+		m_keyType(std::move(_keyType)),
+		m_valueType(std::move(_valueType)),
+		m_keyName(std::move(_keyName)),
+		m_valueName(std::move(_valueName)) 
```
```suggestion
		m_valueName(std::move(_valueName))
```
:)

hrkrshnn replied:
Kinda surprised the style checker didn't catch this. edit: it actually did.

hrkrshnn commented on libsolidity/ast/ASTJsonImporter.cpp:
```diff
@@ -648,7 +648,9 @@ ASTPointer<Mapping> ASTJsonImporter::createMapping(Json::Value const&  _node)
 		_node,
 		convertJsonToASTNode<TypeName>(member(_node, "keyType")),
-		convertJsonToASTNode<TypeName>(member(_node, "valueType"))
+		convertJsonToASTNode<TypeName>(member(_node, "valueType")),
+		convertJsonToASTNode<ASTString>(member(_node, "keyName")),
+		convertJsonToASTNode<ASTString>(member(_node, "valueName"))
```
The equivalent code was missing in `ASTJsonExporter`. This is why the cli test was failing with
```
Looking at 2940 .sol files...
................................................................................................................................................ERROR: AST reimport failed for input file /home/hari/s/r2/test/libsolidity/syntaxTests/array/pop/storage_with_mapping_pop.sol

Compiler stderr:
Failed to import AST: JSON-Node needs to have 'nodeType' and 'id' fields.

Compiler stdout:

```
I'll fix this.

Kinda wish we had serde magic here :(

hrkrshnn replied:
There is also a nasty bug here. `ASTString` is not an ASTNode. So this conversion would fail at runtime.

The correct way to do it would be:
```diff
-		convertJsonToASTNode<ASTString>(member(_node, "keyName")),
-		convertJsonToASTNode<ASTString>(member(_node, "valueName"))
+		memberAsASTString(_node, "keyName"),
+		memberAsASTString(_node, "valueName")
```

hrkrshnn commented on docs/types/mapping-types.rst:
```diff
@@ -64,6 +67,25 @@ contract that returns the value at the specified address.
+.. code-block:: solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity ^0.8.17;
```
```suggestion
    pragma solidity ^0.8.18;
```

hrkrshnn commented on libsolidity/ast/ASTJsonExporter.cpp:
```diff
@@ -599,6 +599,8 @@ bool ASTJsonExporter::visit(Mapping const& _node)
 	setJsonNode(_node, "Mapping", {
 		make_pair("keyType", toJson(_node.keyType())),
 		make_pair("valueType", toJson(_node.valueType())),
+		make_pair("keyName", _node.keyName()),
+		make_pair("valueName", _node.valueName()),
```
I wonder if this change would be considered breaking. cc @cameel @ekpyron

hrkrshnn commented:
Also, a fun test case would be the following
```solidity
type name is uint256;
contract C {
    mapping (name name => uint) public names;
}
```

The corresponding auto-generated function `function names(name name) returns (uint);` The compiler would complain that `Name has to refer to a struct, enum or contract.` which is a bit misleading.

Another example is
```solidity
mapping (uint x => uint x) public something;
```

The equivalent function would be `function f(uint x) returns (uint x)` which wouldn't compile because the name is defined twice in the same context.

I think we need to do some additional sanity checks on the names defined in this way.

nikola-matic commented:
Heya @zemse, I've rebased against latest upstream develop to get the build in working order. Are you still interested in working on this PR, or should we have someone else takeover? Also, for future reference, please create a separate branch in your fork for features instead of working directly on `develop`.

nishant-sachdeva commented on docs/types/mapping-types.rst:
```diff
@@ -64,6 +67,25 @@ contract that returns the value at the specified address.
 ``_allowances`` is an example of a mapping type inside another mapping type.
+
+In the example below, the optional ``KeyName`` and ``ValueName`` are provided for the mapping.
+It does not make affect any contract functionality or bytecode, it only sets the ``name`` field
```
```suggestion
It does not affect any contract functionality or bytecode, it only sets the ``name`` field
```

nishant-sachdeva commented on docs/types/mapping-types.rst:
```diff
@@ -4,12 +4,13 @@
-built-in value type, ``bytes``, ``string``, or any contract or enum type. Other user-defined
-or complex types, such as mappings, structs or array types are not allowed.
-``ValueType`` can be any type, including mappings, arrays and structs.
+Mapping types use the syntax ``mapping(KeyType KeyName => ValueType ValueName)`` and variables of
+mapping type are declared using the syntax ``mapping(KeyType KeyName => ValueType ValueName)
```
```suggestion
Mapping types use the syntax ``mapping(KeyType KeyName? => ValueType ValueName?)`` and variables of
mapping type are declared using the syntax ``mapping(KeyType KeyName? => ValueType ValueName?)
```
Suggested this change to indicate that adding named parameters is optional

nikola-matic replied:
Agreed.

nikola-matic commented on test/libsolidity/SolidityTypes.cpp:
```diff
@@ -97,12 +97,16 @@ BOOST_AUTO_TEST_CASE(storage_layout_mapping)
 		{"second", TypeProvider::mapping(
 			TypeProvider::fromElementaryTypeName("uint8"),
-			TypeProvider::fromElementaryTypeName("uint8")
+			TypeProvider::fromElementaryTypeName("uint8"),
```
These parameters are passed in the wrong order, i.e. they don't match the function declaration, which is why the build is failing.

nikola-matic reviewed (changes requested):
The grammar check is failing, since the grammar you've defined allows for named nested mappings, whereas your implementation does not. Given the other nested mapping tests, it seems like you don't want to allow named nested mappings. I'd be curious to know why, since this would then require a specific rule to handle nested mappings? I'll add it to our agenda for the next language design meeting, and we can discuss further there - if you want to join, it's the [Wednesday 15:00 CET meeting].(https://docs.soliditylang.org/en/latest/contributing.html#team-calls)

In any case, there were a few minor issues that I've fixed, which should now hopefully get the build into a workable state (excluding the failing grammar test).

zemse commented:
> it seems like you don't want to allow named nested mappings

```solidity
mapping(uint hello => mapping(uint world => uint whats) up) guys;

// getter for above mapping
function guys(uint hello, uint world) public returns (uint whats);
```

As you can see the mapping name "up" is not used at all. Since it means nothing, still allowing that code to be compiled will mislead people into expecting it to work. 

> since this would then require a specific rule to handle nested mappings?

If adding grammar for this case is complex then it makes sense to remove this requirement, since it's not really a big deal ig. And actually Hari helped me with the grammar previously, I added this check for named nested mappings after his help while adding additional test cases and was not aware I have to fix the language grammar as well (I'm not familiar with that, will need to figure it out). 

> I'll add it to our agenda for the next language design meeting

Thanks! I've added the call to my calendar.

> Are you still interested in working on this PR, or should we have someone else takeover? 

I am up for working on fixing minor issues in the PR. But if it's faster through someone from the solidity team, then we should do that probably. Specifically for the language grammar fix (due to the named nested mapping names), it would be great if I get some help.

> Also, for future reference, please create a separate branch in your fork for features instead of working directly on develop.

Sure, sorry about directly using the develop branch on my fork. Is there a way to change the fork's branch in this same PR? I think I will have to close this PR and create a new PR but that would lose the discussions in this thread. If that is okay, I can do it. And for future things, I'll make sure to not use default branches on fork.

nikola-matic commented:
I'm far from an expert so someone else will from the team will have to chime in.
>As you can see the mapping name "up" is not used at all. Since it means nothing, still allowing that code to be compiled will mislead people into expecting it to work.
Yup, looks like it. The only question I have is whether it's 'used', or rather, whether it shows up in the ABI or not? The answer to this question will essentially dictate whether we alter the grammar, or the code.
>If adding grammar for this case is complex then it makes sense to remove this requirement, since it's not really a big deal ig. And actually Hari helped me with the grammar previously, I added this check for named nested mappings after his help while adding additional test cases and was not aware I have to fix the language grammar as well (I'm not familiar with that, will need to figure it out).
I'm sure this won't be a problem either way, it's just a matter of making a decision. We can discuss this on Wednesday.
>I am up for working on fixing minor issues in the PR. But if it's faster through someone from the solidity team, then we should do that probably. Specifically for the language grammar fix (due to the named nested mapping names), it would be great if I get some help.
Awesome - as long as you're responsive, it's not really a big deal. The only reason I asked was because the PR has been inactive for quite some time. Whatever decision we make, I'm guessing it won't require a lot of work; we'll help out with whatever you need.
>Sure, sorry about directly using the develop branch on my fork. Is there a way to change the fork's branch in this same PR? I think I will have to close this PR and create a new PR but that would lose the discussions in this thread. If that is okay, I can do it. And for future things, I'll make sure to not use default branches on fork.
No need to do anything regarding this now, I just mentioned it for future reference in case you decide to continue contributing.

chriseth commented:
Just randomly throwing in some important test cases we need to cover:
 - named mapping parameters for storage pointers of mapping types as local variables (and function parameters)
 - named mapping parameters for state variables of function type that takes a mapping as parameter
 - what about mappings as struct members?

Not saying anything about this should or should not work, but we should certainly have tests for these.

nikola-matic commented:
As suggested by @ekpyron in the language design meeting:
1. Allow names everywhere (i.e. remove checks)
2. More tests

zemse commented:
Summary of the changes done:
1. Removed the commit ee0fee7f7613f45439143e42b37ed84ff6236a7b that disallowed param for mapping value types.
2. Added more tests in 4aedd7755ce215995cae080f644fb984f22e4522.
3. Rebased over upstream/develop branch

nikola-matic commented on test/libsolidity/syntaxTests/parsing/mapping_with_names_conflict_3.sol:
```diff
@@ -0,0 +1,7 @@
+}
+// ----
+// DeclarationError 5609: (45-84): Conflicting parameter name "owner" in mapping.
+// DeclarationError 1809: (20-85): Conflicting parameter name "owner" in recursive mapping.
```
~~Why does this one have a different error code, i.e. 1809 vs 5609?~~ Ah OK, it's a different error in case the nested type is a mapping, so you're differentiating between conflicting key -> key and key -> value names. The error message is the same though, which is perfectly fine, but in that case, the errors should have the same code.

zemse replied:
Since both these cases serve the same purpose, it might make sense to just have the same error and code in both cases.

> so you're differentiating between conflicting key -> key and key -> value names

I did that because the error logic was at different places in the code and I tend to write more descriptive error msg if possible. But here I agree the solidity dev that ends up on this error doesn't benefit much.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,57 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+				if (keyName == childKeyName)
+				{
+					m_errorReporter.declarationError(
+						1809_error,
```
```suggestion
						5609_error,
```

nikola-matic commented on test/libsolidity/syntaxTests/parsing/mapping_with_names_struct_member_4.sol:
```diff
@@ -0,0 +1,6 @@
+contract test {
+    struct Person {
+        mapping(uint phone => mapping(uint call => uint time) callTimes) friends;
```
Adding another test case with an additional level of nesting would be nice, especially if it hits the conflicting parameter error between the root and leaf mapping, e.g.
```
mapping(uint phone => mapping(uint call => mapping(uint phone => something) leafMapping) callTimes) friends;
```

zemse replied:
There was a typo in the logic and this test case caught that. I've updated the fix in commit [7438248](https://github.com/ethereum/solidity/pull/13384/commits/743824807e95d2ca282996b50ea14f945a79c9f3).

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,57 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+					m_errorReporter.declarationError(
+						1809_error,
+						_mapping.location(),
+						"Conflicting parameter name \"" + keyName + "\" in recursive mapping."
```
Oh yeah, one thing that kinda irks me is this being called a recursive mapping. Nested would be a more appropriate term here.

zemse replied:
I've just combined the error into single one, removing the "recursive" word itself. So now it looks like: `Conflicting parameter name "nameSame" in mapping.`. Lmk if we have to change this into something better.

nikola-matic commented:
I've pushed an empty commit, squash and pushed in the hopes of restarting the build successfully, but unfortunately we seem to have CI problems, which will hopefully be resolved soon. In fact, as soon as that happens, I'll restart the build to see whether everything passes, and then I'll resume reviewing.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,56 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+	_mapping.annotation().type = TypeProvider::mapping(keyType, keyName, valueType, valueName);
+
+	// Check if parameter names are conflicting.
+	if (keyName != "")
```
```suggestion
	if (!keyName.empty())
```

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,56 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+				ASTString childKeyName = childMappingType->keyName();
+				if (keyName == childKeyName)
+				{
+					m_errorReporter.declarationError(
```
The `chk_errorcodes` step is failing because you can't throw an error with the same code twice, i.e. this one, and the one on line 310; instead you'll have to use a boolean to indicate whether an error needs to be thrown, and then check at the end of the loop iteration and report error if true.

zemse replied:
I've just pushed a change for this.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,59 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+			else
+			{
+				// Compare top mapping's key name with the value name.
+				if (keyName == currentValueName)
+				{
+					isError = true;
+					break;
+				}
```
```suggestion
				if (keyName == currentValueName)
					isError = true;
```

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,59 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+			{
+				// Compare top mapping's key name with child mapping's key name.
+				ASTString childKeyName = childMappingType->keyName();
+				if (keyName == childKeyName)
+				{
+					isError = true;
+					break;
+				}
```
This is semantically different than the previous approach (which was correct), as you're now break-ing from the loop as soon as you encounter an error instead of continuing to type check, which is ultimately why you had to remove one line form [this test](https://github.com/ethereum/solidity/pull/13384/files#diff-789fefb5e21fd40f374366380ff975e1bfdd6f2b2ed9af02a3c6982e9a449e07). All you need to do here is set the boolean indicator, and continue as per usual.

```suggestion
				if (keyName == childKeyName)
					isError = true;
```

zemse replied:
~I will make this change but I still don't see the value since the error would still be reported once. Because in previous logic it was reported every time it was discovered and now only the last error would be reported.~

Okay, this makes sense, after the for-loop update it should be reporting all errors.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,59 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+					break;
+				}
+				break; // We arrived at the end of mapping recursion.
+			}
+		}
+		if (isError)
+		{
+			m_errorReporter.declarationError(
+				1809_error,
+				_mapping.location(),
+				"Conflicting parameter name \"" + keyName + "\" in mapping."
+			);
+		}
```
Error checking should be done within the `for` loop after every iteration, otherwise you'll always miss one case.

```suggestion
			}
			if (isError)
			{
				m_errorReporter.declarationError(
					1809_error,
					_mapping.location(),
					"Conflicting parameter name \"" + keyName + "\" in mapping."
				);
			}
		}
```

nikola-matic commented on test/libsolidity/syntaxTests/parsing/mapping_with_names_struct_member_6.sol:
```diff
@@ -0,0 +1,7 @@
+contract test {
+    struct Person {
+        mapping(uint nameSame => mapping(uint name1 => mapping(uint nameSame => uint name2) name3) name4) name5;
```
Another three-level nested mapping test, but this time valid, please :)

zemse replied:
I've added that, please review.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,63 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+			}
+
+			// If we are at the end of the mapping recursion, we can break.
+			if (!loop) {
```
```suggestion
			if (!loop) 
			{
```

nikola-matic replied:
Wait, do you even need this? If `loop` is false, it will not run the next iteration anyway.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,63 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+		auto childMappingType = dynamic_cast<MappingType const*>(valueType);
+		ASTString currentValueName = valueName;
+		bool loop = true;
+		while (loop) {
```
~~This could be a `do { } while(childMappingType)` loop, which should hopefully make the logic simpler by not requiring the `loop` variable.~~
Tried it, and it of course doesn't work as we want to reach the case where the `childMappingType` is not a mapping, and then perform another iteration.

nikola-matic commented on libsolidity/analysis/DeclarationTypeChecker.cpp:
```diff
@@ -267,14 +267,58 @@ void DeclarationTypeChecker::endVisit(Mapping const& _mapping)
+		auto childMappingType = dynamic_cast<MappingType const*>(valueType);
+		ASTString currentValueName = valueName;
+		bool loop = true;
+		while (loop) {
```
```suggestion
		while (loop)
		{
```

nikola-matic reviewed (approved):
I've addressed the remaining issues, rebased and squashed, and am satisfied with the state of the PR at the moment, and am thus approving. @cameel or @ekpyron, please take over for the final review.
